ahn/training/train_zero3.py

```diff
@@ -0,0 +1,256 @@
+        tokenizer.pad_token_id = tokenizer.eos_token_id
+
+    # 10. Load dataloaders
+    logger.info("Loading train data loaders...")
+    train_data_loader = get_data_loader_hfstyle(config, tokenizer, "train")
+    logger.info("Loading valid data loaders...")
+    valid_data_loader = get_data_loader_hfstyle(config, tokenizer, "valid")
+
+    # 11. Setup gradient checkpointing
+    if config["efficiency"]["gradient_checkpointing"]:
+        model.gradient_checkpointing_enable()
+
+    # 12. Setup cuda and cudnn settings
+    cudnn.benchmark = False
+    if config["efficiency"]["allow_tf32"]:
+        cuda.matmul.allow_tf32 = True
+
+    # 13. Setup deepspeed engine
+    logger.info("Setup deepspeed engine...")
+    engine = deepspeed.initialize(
+        model=model,
+        config=config["deepspeed"],
+        model_parameters=optimized_params(
+            model,
+            config["deepspeed"]["optimizer"]["params"]["weight_decay"],
+        ),
+    )[0]
+
+    # 14. Setup wandb monitoring
+    exp_name = f'{config["training"]["exp_name"]}-{get_timestamp()}'
+    logger.info(f"Experiment name: {exp_name}")
+    if dist.get_rank() == 0:
+        wandb.init(
+            project=config["training"]["project"],
+            name=exp_name,
+        )
+    dist.barrier()
+
+    # 15. Skip training steps if model is already trained
+    if config["training"]["current_step"] > 0:
+        for _ in range(config["training"]["current_step"]):
+            next(train_data_loader)
+
+    table_data = []
+    total_step = (
+        config["training"]["total_step"]
+        if not config.get("debug")
+        else len(train_data_loader)
+    )
+    # 16. Start training
+    while True:
+        if config["training"]["current_step"] >= total_step:
+            break
+        engine.module.train()
+        for i, train_data in enumerate(train_data_loader):
+            if config["training"]["current_step"] >= total_step:
+                break
+            loss = engine(
+                **{
+                    k: v.cuda() if torch.is_tensor(v) else v
+                    for k, v in train_data.items()
+                }
+            ).loss
+
+            if (
+                i % config["training"]["train_print_every"] == 0
+                or i == len(train_data_loader) - 1
+            ):
+                logger.info(
+                    f"[train] EPOCH: {config['training']['current_step']/len(train_data_loader):.3f} "
+                    f"STEP: {i}/{len(train_data_loader) - 1}, LOSS: {loss}"
+                )
+            if dist.get_rank() == 0:
+                wandb.log(
+                    data={
+                        "train_loss": loss.item(),
+                        "train_ppl": math.exp(loss.item()),
+                    },
+                    step=config["training"]["current_step"],
+                )
+
+            engine.backward(loss)
+            engine.step()
+            torch.cuda.empty_cache()
```

Comment:
should we empty gpu cache?

Reply:
I added it becuase the following waring:
```pytorch allocator cache flushes since last step. this happens when there is high memory pressure and is detrimental to performance. if this is happening frequently consider adjusting settings to reduce memory consumption. If you are unable to make the cache flushes go away consider adding get_accelerator().empty_cache() calls in your training loop to ensure that all ranks flush their caches at the same time```
with experiment, I find that there's not much effect on training time but still it does not resolve the warninig above.